lib/Dialect/Arc/Transforms/LowerState.cpp

```diff
@@ -772,28 +786,23 @@ LogicalResult LowerStatePass::runOnModule(HWModuleOp moduleOp,
   LLVM_DEBUG(llvm::dbgs() << "Lowering state in `" << moduleOp.getModuleName()
                           << "`\n");
   ModuleLowering lowering(moduleOp, stats);
+
+  // Add sentinel ops to separate state allocations from clock trees.
+  lowering.stateBuilder.setInsertionPointToStart(moduleOp.getBodyBlock());
+
+  Operation *stateSentinel =
+      lowering.stateBuilder.create<hw::OutputOp>(moduleOp.getLoc());
+  Operation *clockSentinel =
+      lowering.stateBuilder.create<hw::OutputOp>(moduleOp.getLoc());
+
+  lowering.stateBuilder.setInsertionPoint(stateSentinel);
+  lowering.clockBuilder.setInsertionPoint(clockSentinel);
+
   lowering.addStorageArg();
   if (failed(lowering.lowerStates()))
     return failure();
   if (failed(lowering.lowerExtModules(symtbl)))
     return failure();
-
-  // Since we don't yet support derived clocks, simply delete all clock trees
-  // that are not driven by a primary input and emit a warning.
-  for (auto &[clock, clockLowering] : lowering.clockLowerings) {
-    auto clockTreeOp = dyn_cast<ClockTreeOp>(clockLowering->treeOp);
-    if (!clockTreeOp)
-      continue;
-
-    if (clock.isa<BlockArgument>())
-      continue;
-    auto d =
-        mlir::emitWarning(clock.getLoc(), "unsupported derived clock ignored");
-    d.attachNote()
-        << "only clocks through a top-level input are supported at the moment";
-    clockTreeOp.erase();
-  }
-
```

Comment:
So cool that this limitation is gone now! 😎

Reply:
Yeah! It doesn't really work yet because we need a true `eval` function with state update legalization for this to work, but it's a step in that direction :smiley: